### PR TITLE
Igal/mgmt 5323 allow adding costum manifests

### PR DESCRIPTION
**Pull request by tsorya**
We want to allow users to create/delete custom manifests through cluster
    deployment crd.
    User should create configmap with custom manifests as data and add it's name to manifestsConfigMapRef  proviosining field
    This should be done only before cluster starts installation and will be
    ignored after it.

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -876,6 +876,7 @@ paths:
       tags:
         - manifests
       security:
+        - agentAuth: []
```
Comment (rollandf) [inline swagger.yaml:879]: Why is this needed?

File: subsystem/kubeapi_test.go [hunk 2]
```diff
@@ -1277,4 +1304,62 @@ var _ = Describe("[kube-api]cluster installation", func() {
...
+		cluster := getClusterFromDB(ctx, kubeClient, db, clusterKubeName, waitForReconcileTimeout)
+		configureLocalAgentClient(cluster.ID.String())
+		Eventually(func() int {
+			manifests, err := agentBMClient.Manifests.ListClusterManifests(ctx, &manifests.ListClusterManifestsParams{
+				ClusterID: *cluster.ID,
+			})
+			Expect(err).NotTo(HaveOccurred())
+			return len(manifests.Payload)
+		}, "1m", "2s").Should(Equal(1))
+	})
```
Comment (rollandf) [inline subsystem/kubeapi_test.go:1364]: What will happen if the config map is updated after the CD is created?

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 3]
```diff
@@ -1090,8 +1103,272 @@ var _ = Describe("cluster reconcile", func() {
...
+				},
+				PullSecret: testPullSecretVal,
+			}
+
+			mockInstallerInternal.EXPECT().GetClusterByKubeKey(gomock.Any()).Return(backEndCluster, nil)
+			mockManifestsApi.EXPECT().ListClusterManifestsInternal(gomock.Any(), gomock.Any()).Return(nil, nil).Times(1)
+			request := newClusterDeploymentRequest(cluster)
+			cluster = getTestCluster()
+			ref := &corev1.LocalObjectReference{Name: "cluster-install-config"}
+			// data := map[string]string{"test.yaml": "test"}
```
Comment (rollandf) [inline internal/controller/controllers/clusterdeployments_controller_test.go:1255]: delete comment

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 4]
```diff
@@ -1090,8 +1103,272 @@ var _ = Describe("cluster reconcile", func() {
...
+			mockInstallerInternal.EXPECT().GetClusterByKubeKey(gomock.Any()).Return(backEndCluster, nil)
+			mockManifestsApi.EXPECT().ListClusterManifestsInternal(gomock.Any(), gomock.Any()).Return(nil, nil).Times(1)
+			mockManifestsApi.EXPECT().CreateClusterManifestInternal(gomock.Any(), gomock.Any()).Return(nil, errors.Errorf("error")).Times(1)
+			request := newClusterDeploymentRequest(cluster)
+			cluster = getTestCluster()
+			cluster.Spec.Provisioning.ManifestsConfigMapRef = ref
+			Expect(c.Update(ctx, cluster)).Should(BeNil())
+			result, err := cr.Reconcile(ctx, request)
+			Expect(err).To(BeNil())
+			Expect(result).To(Equal(ctrl.Result{RequeueAfter: 10000000000}))
```
Comment (rollandf) [inline internal/controller/controllers/clusterdeployments_controller_test.go:1310]: Expect(result).To(Equal(ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}))

File: swagger.yaml (hunk 1 above)
Comment (tsorya) [reply swagger.yaml:879]: for tests, as i understand rest will be disabled on crd and i need this api to be able to validate subsystem

File: subsystem/kubeapi_test.go (hunk 2 above)
Comment (tsorya) [reply subsystem/kubeapi_test.go:1364]: it will add manifests on next reconcile.

File: swagger.yaml (hunk 1 above)
Comment (rollandf) [reply swagger.yaml:879]: agent auth will not disabled on CRD...
I am not a fan of changing security for tests. Any other way to validate?

File: subsystem/kubeapi_test.go (hunk 2 above)
Comment (rollandf) [reply subsystem/kubeapi_test.go:1364]: Who says there will be another reconcile?

File: swagger.yaml (hunk 1 above)
Comment (eranco74) [reply swagger.yaml:879]: @tsorya the REST API is getting tested in subsystem.

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 5]
```diff
@@ -428,6 +440,74 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
...
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ClusterDeploymentsReconciler) addCustomManifests(ctx context.Context, clusterDeployment *hivev1.ClusterDeployment,
+	cluster *common.Cluster) error {
+
+	if !common.IsInPreInstallationState(cluster) {
```
Comment (eranco74) [inline internal/controller/controllers/clusterdeployments_controller.go:479]: Instead of skipping consider validating that the `alreadyCreatedManifests` are synced with the `ManifestsConfigMapRef`.
If not set a condition to "Failed" with a relevant error.

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 6]
```diff
@@ -428,6 +440,74 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
...
+	configuredManifests := &corev1.ConfigMap{}
+	configuredManifests.Data = map[string]string{}
+	if clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef != nil {
+		configuredManifests = &corev1.ConfigMap{}
+		err = r.Get(ctx, types.NamespacedName{Namespace: clusterDeployment.Namespace, Name: clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef.Name}, configuredManifests)
+		if err != nil {
+			r.Log.WithError(err).Errorf("Failed to get configmap %s in %s", clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef.Name, clusterDeployment.Namespace)
+			return err
+		}
+	}
```
Comment (eranco74) [inline internal/controller/controllers/clusterdeployments_controller.go:506]: Consider extracting to another function and calling that function from `syncManifests` where the `configuredManifests` is actually used.

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 7]
```diff
@@ -428,6 +440,74 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
 	return nil
 }
 
+func (r *ClusterDeploymentsReconciler) syncManifests(ctx context.Context, cluster *common.Cluster, manifests map[string]string, alreadyCreatedManifests models.ListManifests) error {
```
Comment (eranco74) [inline internal/controller/controllers/clusterdeployments_controller.go:443]:
```suggestion
func (r *ClusterDeploymentsReconciler) syncManifests(ctx context.Context, cluster *common.Cluster, manifestsFromConfigMap map[string]string, alreadyCreatedManifests models.ListManifests) error {
```

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 8]
```diff
@@ -428,6 +440,74 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
...
+	for _, manifest := range alreadyCreatedManifests {
+		if _, ok := manifests[manifest.FileName]; !ok {
+			r.Log.Infof("Deleting cluster %s manifest %s", cluster.ID, manifest.FileName)
+			_ = r.Manifests.DeleteClusterManifestInternal(ctx, operations.DeleteClusterManifestParams{
+				ClusterID: *cluster.ID,
+				FileName:  manifest.FileName,
+				Folder:    swag.String(models.ManifestFolderOpenshift),
+			})
+		} else {
+			delete(manifests, manifest.FileName)
```
Comment (eranco74) [inline internal/controller/controllers/clusterdeployments_controller.go:455]: Nice idea, add a comment about why you do it.
// Manifest already exists removing it from the map because later we will call `CreateClusterManifestInternal` for all manifest that are left in the map

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 6 above)
Comment (eranco74) [reply internal/controller/controllers/clusterdeployments_controller.go:506]: I think such validation is required before we allow the installation to start.

File: internal/manifests/manifests.go [hunk 9]
```diff
@@ -118,20 +141,34 @@ func (m *Manifests) ListClusterManifests(ctx context.Context, params operations.
...
 	log := logutil.FromContext(ctx, m.log)
 	log.Infof("Deleting manifest from cluster %s", params.ClusterID)
 
 	cluster, apierr := cluster.GetCluster(ctx, m.log, m.db, params.ClusterID.String())
 	if apierr != nil {
-		return common.GenerateErrorResponder(apierr)
+		return apierr
+	}
+
+	if !common.IsInPreInstallationState(cluster) {
```
Comment (eranco74) [inline internal/manifests/manifests.go:168]: So far we didn't handle it?
:exploding_head:

File: subsystem/kubeapi_test.go (hunk 2 above)
Comment (tsorya) [reply subsystem/kubeapi_test.go:1364]: then user will miss new values from configmap. Not sure we should add watch on configmaps but we can. I will look if there is something like this in hive

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 10]
```diff
@@ -428,6 +440,74 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
 	return nil
 }
 
+func (r *ClusterDeploymentsReconciler) syncManifests(ctx context.Context, cluster *common.Cluster, manifests map[string]string, alreadyCreatedManifests models.ListManifests) error {
+	r.Log.Debugf("Going to sync list of given with already created manifests")
+
+	for _, manifest := range alreadyCreatedManifests {
```
Comment (eranco74) [inline internal/controller/controllers/clusterdeployments_controller.go:446]: Actually, this implementation isn't correct.
You can't just compare the manifest file name.
You should validate the manifest content

**Review by eranco74 — CHANGES_REQUESTED** [review]

File: swagger.yaml (hunk 1 above)
Comment (tsorya) [reply swagger.yaml:879]: @rollandf thats why i added agent auth. 
I don't see any other option to validate it from subsystem to say the truth. If you have idea will be glad to hear

File: subsystem/kubeapi_test.go (hunk 2 above)
Comment (tsorya) [reply subsystem/kubeapi_test.go:1364]: we can open ticket for it and fix it another time. Not sure when this case will be relevant to say the truth and it is relevant for all other configs we are adding

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 5 above)
Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:479]: I don't want to do anything on manifests if we are already installing. We are adding more manifests on installation process and i don't want to sync in that case. In any case it will be too late.

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 10 above)
Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:446]: so didn't wanted to handle this case :(

File: swagger.yaml (hunk 1 above)
Comment (avishayt) [reply swagger.yaml:879]: If the agent is not calling this API don't add agentAuth

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 6 above)
Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:506]: Not sure i understand.  if configmap reference was set and configmap will not be with us, we will fail and return error.
What else can we do?

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 10 above)
Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:446]: It will be very problematic one, not sure if it worth supporting it

Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:446]: maybe will be enough to describe process of changing file content? cause it will require to run file download for each file. I think we shouldn't support it (too expensive)

File: swagger.yaml (hunk 1 above)
Comment (tsorya) [reply swagger.yaml:879]: yeah, knew it :) .
not sure what to do with subsystem though, will we allow user api calls in subsystem? @filanov ?

**Comment by rollandf** [comment]
Please also add a section in documentation and maybe an example.

File: internal/common/common.go [hunk 11]
```diff
@@ -70,3 +71,13 @@ func GetBootstrapHost(cluster *Cluster) *models.Host {
 func IsSingleNodeCluster(cluster *Cluster) bool {
 	return swag.StringValue(cluster.HighAvailabilityMode) == models.ClusterHighAvailabilityModeNone
 }
+
+// IsInPreInstallationState verifies if cluster is in pre-installation stage
+func IsInPreInstallationState(cluster *Cluster) bool {
```
Comment (filanov) [inline internal/common/common.go:76]: Maybe i should be part of cluster API, after all, it's a cluster-specific logic

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 12]
```diff
@@ -428,6 +440,74 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
...
+	if !common.IsInPreInstallationState(cluster) {
+		r.Log.Debugf("Cluster is not in pre installation state, skipping")
+		return nil
+	}
+
+	alreadyCreatedManifests, err := r.Manifests.ListClusterManifestsInternal(ctx, operations.ListClusterManifestsParams{
+		ClusterID: *cluster.ID,
+	})
+	if err != nil {
+		r.Log.WithError(err).Errorf("Failed to list cluster %s manifests", cluster.ID)
```
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:488]: Cluster ID doesn't mean anything for the user.

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 13]
```diff
@@ -428,6 +440,74 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
...
+	if clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef == nil && len(alreadyCreatedManifests) < 1 {
+		r.Log.Debugf("Nothing to do, skipping manifest creation")
+		return nil
+	}
+
+	configuredManifests := &corev1.ConfigMap{}
+	configuredManifests.Data = map[string]string{}
+	if clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef != nil {
+		configuredManifests = &corev1.ConfigMap{}
+		err = r.Get(ctx, types.NamespacedName{Namespace: clusterDeployment.Namespace, Name: clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef.Name}, configuredManifests)
```
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:501]: please break long lines.

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 14]
```diff
@@ -1090,8 +1103,272 @@ var _ = Describe("cluster reconcile", func() {
 			Expect(result).To(Equal(ctrl.Result{}))
 		})
 
+		It("Update manifests - skip cases", func() {
+			backEndCluster := &common.Cluster{
```
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller_test.go:1107]: isn't there a lot of common code here? maybe it should go into a context with a single before each that creates the cluster?

File: internal/common/common.go (hunk 11 above)
Comment (filanov) [reply internal/common/common.go:84]: Why was it resovled?

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 15]
```diff
@@ -70,9 +73,11 @@ type ClusterDeploymentsReconciler struct {
 	ClusterApi       cluster.API
 	HostApi          host.API
 	CRDEventsHandler CRDEventsHandler
+	Manifests        manifests.ClusterManifestsInternals
 }
 
 // +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;update;create
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch
```
[the remark this answers is not in the document]
Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:80]: will check

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 16]
```diff
@@ -426,6 +443,91 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
...
+	if err != nil {
+		return err
+	}
+
+	// delete all manifests that are not part of configmap
+	// skip errors
+	for _, manifest := range alreadyCreatedManifests {
+		if _, ok := manifestsFromConfigMap[manifest.FileName]; !ok {
+			r.Log.Infof("Deleting cluster deployment %s manifest %s", cluster.KubeKeyName, manifest.FileName)
+			_ = r.Manifests.DeleteClusterManifestInternal(ctx, operations.DeleteClusterManifestParams{
```
[the remark this answers is not in the document]
Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:461]: yep, i think we shouldn't look on it

File: subsystem/kubeapi_test.go [hunk 17]
```diff
@@ -1351,4 +1378,62 @@ var _ = Describe("[kube-api]cluster installation", func() {
...
+			Namespace: Options.Namespace,
+			Name:      host.ID.String(),
+		}
+		By("Approve Agent")
+		Eventually(func() error {
+			agent := getAgentCRD(ctx, kubeClient, key)
+			agent.Spec.Approved = true
+			return kubeClient.Update(ctx, agent)
+		}, "30s", "10s").Should(BeNil())
+		checkClusterCondition(ctx, clusterKey, controllers.ClusterReadyForInstallationCondition, controllers.ClusterReadyReason)
```
[the remark this answers is not in the document]
Comment (tsorya) [reply subsystem/kubeapi_test.go:1417]: @rollandf i didn't see it started and why it should? if we fail in middle of reconcile we will not proceed to install command.
But maybe worth looking on conditions too

File: docs/Kube-API.md [hunk 18]
```diff
@@ -220,3 +220,42 @@ metadata:
...
+metadata:
+  name: my-baremetal-cluster
+  namespace: mynamespace
+  annotations:
+    hive.openshift.io/try-install-once: "true"
+spec:
+    manifestsConfigMapRef:
+      name: my-baremetal-cluster-install-manifests
+```
+If manifests provided in configmap data section will be in bad format or configmap will not exists but will be referenced
```
[the remark this answers is not in the document]
Comment (tsorya) [reply docs/Kube-API.md:260]: why? this is right behavior? not sure we need to provide in doc that it will requeue after a minute. Implementation details

File: subsystem/kubeapi_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (rollandf) [reply subsystem/kubeapi_test.go:1417]: mmh, I guess you are right. The failure will exit the func.

File: docs/Kube-API.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (rollandf) [reply docs/Kube-API.md:260]: I meant that the error in the Sync condition will be available only if the Cluster is able to install.

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:80]: yes, we need

File: internal/common/common.go (hunk 11 above)
Comment (tsorya) [reply internal/common/common.go:84]: i think i missed it after pushing fixes and closing all previous. 
In any case i don't think it should be api call, it will be overkill.

**Review by eranco74 — APPROVED** [review]

File: internal/common/common.go (hunk 11 above)
Comment (eranco74) [reply internal/common/common.go:84]: I don't think @filanov means it should be an API call just move this function to be under clusterapi package instead of common

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 19]
```diff
@@ -426,6 +443,91 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
...
+		}
+	}
+	return configuredManifests.Data, nil
+}
+
+func (r *ClusterDeploymentsReconciler) addCustomManifests(ctx context.Context, clusterDeployment *hivev1.ClusterDeployment,
+	cluster *common.Cluster) error {
+
+	// added just in case we it will run after cluster will be in the middle of installation
+	if !common.IsInPreInstallationState(cluster) {
```
Comment (eranco74) [inline internal/controller/controllers/clusterdeployments_controller.go:507]: This seems redundent, this method is called only when:
``` ready, err := r.isReadyForInstallation(ctx, clusterDeployment, cluster)```
returns ready=True

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 20]
```diff
@@ -426,6 +443,91 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
...
+	if err != nil {
+		r.Log.WithError(err).Errorf("Failed to list manifests for %q cluster deployment", clusterDeployment.Name)
+		return err
+	}
+
+	// if reference to manifests was deleted from cluster deployment
+	// but we already added some in previous reconcile loop, we want to clean them.
+	// if no reference were provided we will delete all manifests that were in the list
+	if clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef == nil && len(alreadyCreatedManifests) < 1 {
+		r.Log.Debugf("Nothing to do, skipping manifest creation")
```
Comment (eranco74) [inline internal/controller/controllers/clusterdeployments_controller.go:524]: Are you expecting -3 `alreadyCreatedManifests`? why not check if `len(alreadyCreatedManifests) == 0`?

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (eranco74) [reply internal/controller/controllers/clusterdeployments_controller.go:461]: Why not log it?

File: internal/common/common.go (hunk 11 above)
Comment (filanov) [reply internal/common/common.go:84]: Yes, there is no reason to keep this logic inside common functions it's a cluster specific function.

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:461]: it will be logged inside DeleteClusterManifestInternal, no need to do it twice

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 19 above)
Comment (tsorya) [reply internal/controller/controllers/clusterdeployments_controller.go:507]: yeah, sounds like

File: internal/common/common.go (hunk 11 above)
Comment (tsorya) [reply internal/common/common.go:84]: the problem that this will create circle dependency. Cluster api requires manifests api and then manifests api will require cluster. 
I think this sort of functions doesn't need to be api or needs to be some another common interface.
I still prefer to leave it here.

Comment (filanov) [reply internal/common/common.go:84]: Common is not a garbage collector and the controller should call for inventory internal API.
I don't think that this function belongs to common.

Comment (filanov) [reply internal/common/common.go:84]: Instead of having status all over the code, I think that each component has its own functions and keeps the state in this function, you don't need to import cluster API, you can add another function to `InstallerInternals` that will call to `IsInPreInstallationState` of cluster API.

Comment (tsorya) [reply internal/common/common.go:84]: i think we started to over complicate stuff. Really prefer just to remove it and same stuff we did till now by validating list of allowed statuses. As i said what you suggest will cause circle dependency between manifests and clusterApi. I don't want to handle it in this PR for sure

Comment (filanov) [reply internal/common/common.go:84]: IMO it's better than have a spaghetti code
